src/json/mod.rs

```diff
@@ -24,63 +31,88 @@ pub fn create_json(host: &AnalysisHost, crate_name: &str) -> Result<String> {
-            });
-        defs.extend(child_defs);
-        defs
-    }
+    // create the main `Document`
+    let mut document = Document::new()
+        .ty(String::from("crate"))
+        .id(crate_name.to_string())
+        .attributes(String::from("docs"), root_def.docs);
 
-    let mut included: Vec<Document> = Vec::new();
-    let mut relationships: HashMap<String, Vec<Data>> = HashMap::with_capacity(METADATA_SIZE);
+    // Now that we have that, it's time to get the children; these are
+    // the top-level items for the crate.
+    let ids = host.for_each_child_def(root_id, |id, _def| id).unwrap();
+
+
+    // Now, we push all of those children onto a channel. The channel functions
+    // as a work queue; we take an item off, process it, and then if it has
+    // children, push them onto the queue. When the queue is empty, we've processed
+    // everything.
+    //
+    // Additionally, we generate relationships between the crate itself and
+    // these ids, as they're at the top level and hence linked with the crate.
 
-    for def in recur(&root_id, host) {
-        let (ty, relations_key) = match def.kind {
+    let (sender, receiver) = mpsc::channel();
+
+    for id in ids {
+        sender.send(id).unwrap();
+
+        let def = host.get_def(id).unwrap();
+
+        let (ty, child_ty) = match def.kind {
             DefKind::Mod => (String::from("module"), String::from("modules")),
             DefKind::Struct => (String::from("struct"), String::from("structs")),
             _ => continue,
         };
 
+        let data = Data::new().ty(ty.clone()).id(def.qualname.clone());
+
+        document.add_relationship(child_ty, data);
+    }
+
+    // the loop below is basically creating this vector
+    let mut included: Vec<Document> = Vec::new();
+
+    while let Ok(id) = receiver.try_recv() {
+        // push each child to be processed itself, and also record
+        // their ids so we can create the relationships for later
+        let child_ids = host.for_each_child_def(id, |id, _def| {
+            sender.send(id).unwrap();
+            id
+        })?;
+
+        // process this one
+
+        // question: we could do this by cloning it in the call to for_each_child_def
+        // above/below; is that cheaper, or is this cheaper?
+        let def = host.get_def(id).unwrap();
+
+
         // Using the item's metadata we create a new `Document` type to be put in the eventual
         // serialized JSON.
-        included.push(
-            Document::new()
-                .ty(ty.clone())
-                .id(def.qualname.clone())
-                .attributes(String::from("name"), def.name)
-                .attributes(String::from("docs"), def.docs),
-        );
-
-        let item_relationships = relationships.entry(relations_key).or_insert_with(
-            Default::default,
-        );
-        item_relationships.push(Data::new().ty(ty).id(def.qualname));
-    }
+        let (ty, child_ty) = match def.kind {
```

Comment:
Would it be possible to have `String` consts at all? We have a lot of these lying around.

Reply:
Yeah so this is a good question; the second set also differs slightly.

Also, these may want to be literals and call the .to_string below instead, idk

Reply:
We have a lot of `String`s in use in other places who's only thing is to be added as a key. It'd probably be faster to have consts for use throughout. Not a huge thing and can be addressed later. This isn't a blocker.